### DISCUSSION TITLE
Admin 2FA TOTP (auth app) integration example, much more secure than email and suprisingly easy.

### DISCUSSION THREAD

**Discussion by hayden-t** [opencart/opencart#14182] [Ideas]
So recently had a admin account login 'leaked' and used by a 'hacker' to deface site. It made me realise once you get multiple admin users this sort of thing is inevitable. So I looked into 2FA options, as if an intruder has access to a users email account, 2fa via email is no use. OC 4 has 2 step by email as an option, and it was very simple too convert it to use an authy/google authenticator TOTP app instead.
This could be made into a admin switch so one can choose between email or TOTP 2FA.

If you want to set this up (I run OC 4.0.1.1, but you can likely adapt it to other subversions easy), you need a empty folder named totp in your home dir (one up from your public_html) to keep the composer libs. (If this was made integrated into oc officially the internal composer install can be used, but i find its easier to install separate composer instances for each addon i make to avoid composer trying to upgrade the internal oc one)

Then within totp folder run `composer require pragmarx/google2fa bacon/bacon-qr-code`

And then in your DB admin run this SQL:
```
ALTER TABLE `oc_user_login` ADD `expiry` DATETIME NOT NULL AFTER `date_added`; 
ALTER TABLE `oc_user_login` ADD `secret` VARCHAR(20) NOT NULL AFTER `token`; 
```
_i generate a seperate secret to the auth cookie token and store in db, but is this necessary ? would it be any less secure to just make the secret the token and not need an extra db field ? (keeping in mind the token is stored in the client side cookie, but either way if you have either you are accepted by 2fa?)_

And here is the vqmod xml: (the twig goes in admin/view/template/common)
(see next post for updated version)

Then once you enable 2 step in the admin store settings you will be greeted with this prompt :
![image](https://github.com/user-attachments/assets/2d5b597e-d971-4db7-ab58-79b9fb66593b)

If you need to reset a users 2FA, so they get the qr code again remove all logins from user/edit/logins form.

**Comment by hayden-t** [comment]
I realised a bug with the first version with how it handled same user logins across devices, this has been fixed to share the qr secret between device sessions, so that it will work across devices with one 2fa app entry.
[admin_totp_v2.zip](https://github.com/user-attachments/files/18563782/admin_totp_v2.zip)

**Comment by danielkerr** [comment]
l wll add t to the repository

**Reply by hayden-t** [reply]
Hi, its not highly polished or finished, but i think it would be a handy free option to have
